src/ocaml/typing/env.ml

```diff
@@ -496,6 +496,28 @@ type type_descriptions = type_descr_kind
 
 let in_signature_flag = 0x01
 
+let stamped_changes =
+  s_table Stamped_hashtable.create_changes ()
+
+let stamped_add table path value =
+  let rec path_stamp = function
+    | Pident id -> Ident.stamp id
+    | Pdot (t, _) -> path_stamp t
+    | Papply (t1, t2) -> Int.max (path_stamp t1) (path_stamp t2)
+  in
+  let stamp = path_stamp path in
+  let stamp = if stamp = 0 then None else Some stamp in
+  Stamped_hashtable.add table ?stamp path value
+
+let stamped_mem table path =
+  Stamped_hashtable.mem table path
+
+let stamped_find table path =
+  Stamped_hashtable.find table path
+
+let stamped_create n =
+  Stamped_hashtable.create !stamped_changes n
```

Comment:
I guess the answer to my previous comment is: "yes but we need to distinguish between the records made when typing different buffers, this way they will each have their own set of changes from their local store"

Reply:
Not sure which comment you are referring to :)?

Reply:
I was referring to https://github.com/ocaml/merlin/pull/1609#discussion_r1200284145

Reply:
Ah yes. Indeed, we have to keep track of the changes for each buffer independently.
Though the hashtables are already buffer local, so if we could find a way to quickly iterate them, the changelog would be only needed for perf purpose.